rust/pattern/constraint/thing_constraint.rs

```diff
@@ -0,0 +1,433 @@
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ValueConstraint {
+    pub predicate: Predicate,
+    pub value: Value,
+}
+
+impl ValueConstraint {
+    pub fn into_constraint(self) -> Constraint {
+        self.into_thing_constraint().into_constraint()
+    }
+
+    pub fn into_thing_constraint(self) -> ThingConstraint {
+        ThingConstraint::Value(self)
+    }
+
+    pub fn new(predicate: Predicate, value: Value) -> ValueConstraint {
+        if predicate.is_substring() && !matches!(value, Value::String(_)) {
+            panic!("");
+        }
+        ValueConstraint { predicate, value }
+    }
+}
+
+impl Display for ValueConstraint {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.predicate == Predicate::Eq && !matches!(self.value, Value::Variable(_)) {
+            write!(f, "{}", self.value)
+        } else {
+            write!(f, "{} {}", self.predicate, self.value)
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Predicate {
+    // equality
+    Eq,
+    Neq,
+    Gt,
+    Gte,
+    Lt,
+    Lte,
+    // substring
+    Contains,
+    Like,
+}
+
+impl Predicate {
+    pub fn is_equality(&self) -> bool {
+        use Predicate::*;
+        matches!(self, Eq | Neq | Gt | Gte | Lt | Lte)
+    }
+
+    pub fn is_substring(&self) -> bool {
+        use Predicate::*;
+        matches!(self, Contains | Like)
+    }
+}
+
+impl From<String> for Predicate {
+    fn from(string: String) -> Self {
+        use Predicate::*;
+        match string.as_str() {
+            "=" => Eq,
+            _ => todo!(),
+        }
+    }
+}
+
+impl Display for Predicate {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use Predicate::*;
+        write!(
+            f,
+            "{}",
+            match self {
+                Eq => "=",
+                Neq => "!=",
+                _ => todo!(),
+            }
+        )
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Value {
+    Long(i64),
+    Double(f64),
+    Boolean(bool),
+    String(String),
+    DateTime(NaiveDateTime),
+    Variable(Box<ThingVariable>),
+}
+impl Eq for Value {} // can't derive, because f32 does not implement Eq
+
+impl From<i64> for Value {
+    fn from(int: i64) -> Value {
+        Value::Long(int)
+    }
+}
+impl From<&str> for Value {
+    fn from(string: &str) -> Value {
+        Value::String(String::from(string))
+    }
+}
+impl From<String> for Value {
+    fn from(string: String) -> Value {
+        Value::String(string)
+    }
+}
+
+impl TryFrom<NaiveDateTime> for Value {
+    type Error = ErrorMessage;
+
+    fn try_from(date_time: NaiveDateTime) -> Result<Value, ErrorMessage> {
+        if date_time.nanosecond() % 1000000 > 0 {
+            return Err(
+                INVALID_CONSTRAINT_DATETIME_PRECISION.format(&[date_time.to_string().as_str()])
+            );
+        }
+        Ok(Value::DateTime(date_time))
+    }
+}
```

Comment:
DateTime and Boolean are the only attribute value types that don't exactly correspond to the type they are constructed from. Thankfully, `From` implies `TryFrom` for other types.